Assets/Scripts/ProceduralGeneration.cs

```diff
@@ -0,0 +1,192 @@
+            entityIndexToId.Add(index, moveId);
+        }
+        entityLookup.Remove(id);
+        cullGroup.SetBoundingSphereCount(EntityCount);
+    }
+
+    private IEnumerator GenerateAsteroidField()
+    {
+        for (var x = 0; x < numChunksInLevel.x; x++)
+        {
+            for (var y = 0; y < numChunksInLevel.y; y++)
+            {
+                for (var z = 0; z < numChunksInLevel.z; z++)
+                {
+                    var chunkStart = new Vector3Int(x, y, z) * chunkSize;
+                    GenerateChunk(chunkStart);
+                    yield return null;
+                }
+            }
+        }
+        FinishedGenerating = true;
+        Debug.Log("Finished Generating!!!");
+    }
+
+    private void GenerateChunk(Vector3Int chunkStart)
+    {
+        for (var x = 0; x < numUnitCubesInChunk.x; x++)
+        {
+            for (var y = 0; y < numUnitCubesInChunk.y; y++)
+            {
+                for (var z = 0; z < numUnitCubesInChunk.z; z++)
+                {
+                    var unitCubeOffset = chunkStart - Vector3.one * (chunkSize / 2 - unitCubeSize / 2);
+                    var unitCubeCenter = (new Vector3Int(x, y, z) * unitCubeSize) + unitCubeOffset;
+                    GenerateChunkAsteroids(unitCubeCenter);
+                }
+            }
+        }
+    }
+    
+    private void GenerateChunkAsteroids(Vector3 unitCubeCenter)
+    {
+        if (!(Random.value < asteroidDensity)) return;
+        
+        var asteroidSize = Random.Range(asteroidSizeRange[0], asteroidSizeRange[1]);
+        var placementVariationRange = (unitCubeSize - asteroidSize) / 2f;
+        var placementVariation = new Vector3(
+            Random.Range(-placementVariationRange, placementVariationRange),
+            Random.Range(-placementVariationRange, placementVariationRange),
+            Random.Range(-placementVariationRange, placementVariationRange)
+        );
+
+        var asteroidPosition = unitCubeCenter + placementVariation;
+        var asteroid = Instantiate(asteroidPrefab, asteroidPosition, Quaternion.identity, asteroidsCollection.transform);
+        asteroid.transform.localScale *= asteroidSize;
+        asteroid.SetActive(false);
+        AddEntity(asteroid);
+    }
+
+    private void CullingStateChanged(CullingGroupEvent evt)
+    {
+        var entity = entityLookup[entityIndexToId[evt.index]];
+        if (evt.currentDistance <= 0 && !entity.activeInHierarchy)
+        {
+            entity.SetActive(true);
+        }
+        else if (evt.currentDistance > 0 && entity.activeInHierarchy)
+        {
+            entity.SetActive(false);
+        }
+    }
+
+    private IEnumerator UpdateCulledObjectBounds()
+    {
+        while (true)
+        {
+            for (var i = 0; i < EntityCount; i++)
+            {
+                bounds[i].position = entityTransforms[entityIndexToId[i]].position;
+                if (i % 25 == 0)
+                {
+                    yield return null;
```

Comment:
What does this do and why are you doing it every 25 entities?

Reply:
25 entities seemed a decent batch size so as to not clog each frame update.